src/main/java/com/google/sps/servlets/MealServlet.java

```diff
@@ -105,29 +116,56 @@ private void getMealById(HttpServletRequest request, HttpServletResponse respons
             response.sendError(HttpServletResponse.SC_NOT_FOUND);
             return;
         }
-
-        Meal meal = dishes.get(id);
-
-        if (meal != null) {
-            String gson = new Gson().toJson(meal);
-            response.setContentType("application/json");
-            response.getWriter().println(gson);
-            return;
+        Filter propertyFilter = new FilterPredicate("id", FilterOperator.EQUAL, id);
+        Query query = new Query("Meal").setFilter(propertyFilter);
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        PreparedQuery results = datastore.prepare(query);
+        ArrayList<Meal> meals = getDataFromDatastore(results); 
+        
+        if (meals.size() > 1) {
+            response.sendError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+        }
+        if (meals.size() == 1) {
+            Meal meal = meals.get(0);
+            if (meal != null) {
+                String gson = new Gson().toJson(meal);
+                response.setContentType("application/json");
+                response.getWriter().println(gson);
+                return;
+            }
         }
-
         response.sendError(HttpServletResponse.SC_NOT_FOUND);
         return;
     }
 
-     private void returnIdOfSimilar(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    private ArrayList<Meal> getDataFromDatastore(PreparedQuery query) {
+        ArrayList<Meal> result = new ArrayList<>();
+        for (Entity entity : query.asIterable()) {
+            Long id = (Long)entity.getProperty("id");
+            String title = (String) entity.getProperty("title");
+            String description = (String) entity.getProperty("description");
+            ArrayList<String> ingredients = (ArrayList<String>) entity.getProperty("ingredients");
+            String type = (String) entity.getProperty("type");
+
+            Meal meal = new Meal(id, title, description, ingredients, type);
+            result.add(meal);
+        }
+        return result;
+    }
+
+    private void returnIdOfSimilar(HttpServletRequest request, HttpServletResponse response) throws IOException {
         // Fuction redirect to random page
         // TODO(grenlayk): implement suggestions algorithm here for Product Alpha
-        Random rand = new Random(); 
-        int randomId = rand.nextInt(dishes.size()); 
+        Query query = new Query("Meal").addSort("id", SortDirection.DESCENDING);
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        PreparedQuery results = datastore.prepare(query);
         
-        String gson = new Gson().toJson(randomId);
-        response.setContentType("application/json");
-        response.getWriter().println(gson);
+        Random rand = new Random(); 
+        int randomId = rand.nextInt(getDataFromDatastore(results).size());
```

Comment:
Id is not necessarily in the range from 0 to "number of entries in database - 1" (i.e. N-1).

For a real-live database, you have to assume that some elements will be missing from the id sequence - usually because a dataset has a long history, and - for example - some elements had to be manually deleted, and in such case no one will redo all the ids. You can't assume what the ids are.

In this case, once you pick a random number X from 0 to N-1, you should read Xth element from the results and get its actual ID - then you'll be sure you have a valid ID.

Also please create a test for this behavior.

Reply:
In "actual ID" you mean id of `Meal` instance (property this.id) or key of entity?